<launch>
    <arg name="is_4dof" default="false" />
    <arg name="solver_type" default="arock" />
    <arg name="g2o_path" default="" />
    <arg name="output_path" default="" />
    <arg name="wait_for_start" default="true" />
    <arg name="ceres_max_solver_time" default="0.5" />
    <arg name="max_steps" default="10" />
    <arg name="agent_num" default="2" />
    <arg name="delay_launch" default="5.0" />
    <arg name="rho_frame_T" default="0.006" />
    <arg name="rho_frame_theta" default="0.00001" />
    <arg name="rho_rot_mat" default="0.16" />
    <arg name="eta_k" default="0.9" />
    <arg name="rviz" default="false" />
    <arg name="rot_init_enable_gravity_prior" default="false" />
    <arg name="debug_rot_init_only" default="false" />
    <arg name="enable_rot_init" default="true" />
    <arg name="enable_linear_pose6d_solver" default="false" />
    <arg name="ignore_infor" default="false" />
    <arg name="simulate_delay_ms" default="0.0" />
    <arg name="max_solving_time" default="10.0" />

    <include file="$(find d2pgo)/launch/d2pgo_test_single.launch" if="$(eval arg('agent_num') > 0)" >
        <arg name="self_id" value="0" />
        <arg name="is_4dof" value="$(arg is_4dof)" />
        <arg name="solver_type" value="$(arg solver_type)" />
        <arg name="g2o_path" value="$(arg g2o_path)/0.g2o" />
        <arg name="output_path" value="$(arg output_path)/0.g2o" />
        <arg name="wait_for_start" value="$(arg wait_for_start)" />
        <arg name="ceres_max_solver_time" value="$(arg ceres_max_solver_time)" />
        <arg name="max_steps" value="$(arg max_steps)" />
        <arg name="rho_frame_T" value="$(arg rho_frame_T)" />
        <arg name="rho_frame_theta" value="$(arg rho_frame_theta)" />
        <arg name="rho_rot_mat" value="$(arg rho_rot_mat)" />
        <arg name="enable_rot_init" value="$(arg enable_rot_init)" />
        <arg name="rot_init_enable_gravity_prior" value="$(arg rot_init_enable_gravity_prior)" />
        <arg name="debug_rot_init_only" value="$(arg debug_rot_init_only)" />
        <arg name="eta_k" value="$(arg eta_k)" />
        <arg name="is_multi" value="true" />
        <arg name="drone_num" value="$(arg agent_num)" />
        <arg name="enable_linear_pose6d_solver" value="$(arg enable_linear_pose6d_solver)" />
        <arg name="ignore_infor" value="$(arg ignore_infor)" />
        <arg name="simulate_delay_ms" value="$(arg simulate_delay_ms)" />
        <arg name="max_solving_time" value="$(arg max_solving_time)" />
    </include>

    <include file="$(find d2pgo)/launch/d2pgo_test_single.launch" if="$(eval arg('agent_num') > 1)" >
        <arg name="self_id" value="1" />
        <arg name="is_4dof" value="$(arg is_4dof)" />
        <arg name="solver_type" value="$(arg solver_type)" />
        <arg name="g2o_path" value="$(arg g2o_path)/1.g2o" />
        <arg name="output_path" value="$(arg output_path)/1.g2o" />
        <arg name="wait_for_start" value="$(arg wait_for_start)" />
        <arg name="ceres_max_solver_time" value="$(arg ceres_max_solver_time)" />
        <arg name="max_steps" value="$(arg max_steps)" />
        <arg name="rho_frame_T" value="$(arg rho_frame_T)" />
        <arg name="rho_frame_theta" value="$(arg rho_frame_theta)" />
        <arg name="rho_rot_mat" value="$(arg rho_rot_mat)" />
        <arg name="enable_rot_init" value="$(arg enable_rot_init)" />
        <arg name="rot_init_enable_gravity_prior" value="$(arg rot_init_enable_gravity_prior)" />
        <arg name="debug_rot_init_only" value="$(arg debug_rot_init_only)" />
        <arg name="eta_k" value="$(arg eta_k)" />
        <arg name="is_multi" value="true" />
        <arg name="drone_num" value="$(arg agent_num)" />
        <arg name="enable_linear_pose6d_solver" value="$(arg enable_linear_pose6d_solver)" />
        <arg name="ignore_infor" value="$(arg ignore_infor)" />
        <arg name="simulate_delay_ms" value="$(arg simulate_delay_ms)" />
        <arg name="max_solving_time" value="$(arg max_solving_time)" />
    </include>

    <include file="$(find d2pgo)/launch/d2pgo_test_single.launch" if="$(eval arg('agent_num') > 2)" >
        <arg name="self_id" value="2" />
        <arg name="is_4dof" value="$(arg is_4dof)" />
        <arg name="solver_type" value="$(arg solver_type)" />
        <arg name="g2o_path" value="$(arg g2o_path)/2.g2o" />
        <arg name="output_path" value="$(arg output_path)/2.g2o" />
        <arg name="wait_for_start" value="$(arg wait_for_start)" />
        <arg name="ceres_max_solver_time" value="$(arg ceres_max_solver_time)" />
        <arg name="max_steps" value="$(arg max_steps)" />
        <arg name="rho_frame_T" value="$(arg rho_frame_T)" />
        <arg name="rho_frame_theta" value="$(arg rho_frame_theta)" />
        <arg name="rho_rot_mat" value="$(arg rho_rot_mat)" />
        <arg name="enable_rot_init" value="$(arg enable_rot_init)" />
        <arg name="rot_init_enable_gravity_prior" value="$(arg rot_init_enable_gravity_prior)" />
        <arg name="debug_rot_init_only" value="$(arg debug_rot_init_only)" />
        <arg name="eta_k" value="$(arg eta_k)" />
        <arg name="is_multi" value="true" />
        <arg name="drone_num" value="$(arg agent_num)" />
        <arg name="enable_linear_pose6d_solver" value="$(arg enable_linear_pose6d_solver)" />
        <arg name="ignore_infor" value="$(arg ignore_infor)" />
        <arg name="simulate_delay_ms" value="$(arg simulate_delay_ms)" />
        <arg name="max_solving_time" value="$(arg max_solving_time)" />
    </include>

    <include file="$(find d2pgo)/launch/d2pgo_test_single.launch" if="$(eval arg('agent_num') > 3)" >
        <arg name="self_id" value="3" />
        <arg name="is_4dof" value="$(arg is_4dof)" />
        <arg name="solver_type" value="$(arg solver_type)" />
        <arg name="g2o_path" value="$(arg g2o_path)/3.g2o" />
        <arg name="output_path" value="$(arg output_path)/3.g2o" />
        <arg name="wait_for_start" value="$(arg wait_for_start)" />
        <arg name="ceres_max_solver_time" value="$(arg ceres_max_solver_time)" />
        <arg name="max_steps" value="$(arg max_steps)" />
        <arg name="rho_frame_T" value="$(arg rho_frame_T)" />
        <arg name="rho_frame_theta" value="$(arg rho_frame_theta)" />
        <arg name="rho_rot_mat" value="$(arg rho_rot_mat)" />
        <arg name="enable_rot_init" value="$(arg enable_rot_init)" />
        <arg name="rot_init_enable_gravity_prior" value="$(arg rot_init_enable_gravity_prior)" />
        <arg name="debug_rot_init_only" value="$(arg debug_rot_init_only)" />
        <arg name="eta_k" value="$(arg eta_k)" />
        <arg name="is_multi" value="true" />
        <arg name="drone_num" value="$(arg agent_num)" />
        <arg name="enable_linear_pose6d_solver" value="$(arg enable_linear_pose6d_solver)" />
        <arg name="ignore_infor" value="$(arg ignore_infor)" />
        <arg name="simulate_delay_ms" value="$(arg simulate_delay_ms)" />
        <arg name="max_solving_time" value="$(arg max_solving_time)" />
    </include>

    <include file="$(find d2pgo)/launch/d2pgo_test_single.launch" if="$(eval arg('agent_num') > 4)" >
        <arg name="self_id" value="4" />
        <arg name="is_4dof" value="$(arg is_4dof)" />
        <arg name="solver_type" value="$(arg solver_type)" />
        <arg name="g2o_path" value="$(arg g2o_path)/4.g2o" />
        <arg name="output_path" value="$(arg output_path)/4.g2o" />
        <arg name="wait_for_start" value="$(arg wait_for_start)" />
        <arg name="ceres_max_solver_time" value="$(arg ceres_max_solver_time)" />
        <arg name="max_steps" value="$(arg max_steps)" />
        <arg name="rho_frame_T" value="$(arg rho_frame_T)" />
        <arg name="rho_frame_theta" value="$(arg rho_frame_theta)" />
        <arg name="rho_rot_mat" value="$(arg rho_rot_mat)" />
        <arg name="enable_rot_init" value="$(arg enable_rot_init)" />
        <arg name="rot_init_enable_gravity_prior" value="$(arg rot_init_enable_gravity_prior)" />
        <arg name="debug_rot_init_only" value="$(arg debug_rot_init_only)" />
        <arg name="eta_k" value="$(arg eta_k)" />
        <arg name="is_multi" value="true" />
        <arg name="drone_num" value="$(arg agent_num)" />
        <arg name="enable_linear_pose6d_solver" value="$(arg enable_linear_pose6d_solver)" />
        <arg name="ignore_infor" value="$(arg ignore_infor)" />
        <arg name="simulate_delay_ms" value="$(arg simulate_delay_ms)" />
        <arg name="max_solving_time" value="$(arg max_solving_time)" />
    </include>

    <include file="$(find d2pgo)/launch/d2pgo_test_single.launch" if="$(eval arg('agent_num') > 5)" >
        <arg name="self_id" value="5" />
        <arg name="is_4dof" value="$(arg is_4dof)" />
        <arg name="solver_type" value="$(arg solver_type)" />
        <arg name="g2o_path" value="$(arg g2o_path)/5.g2o" />
        <arg name="output_path" value="$(arg output_path)/5.g2o" />
        <arg name="wait_for_start" value="$(arg wait_for_start)" />
        <arg name="ceres_max_solver_time" value="$(arg ceres_max_solver_time)" />
        <arg name="max_steps" value="$(arg max_steps)" />
        <arg name="rho_frame_T" value="$(arg rho_frame_T)" />
        <arg name="rho_frame_theta" value="$(arg rho_frame_theta)" />
        <arg name="rho_rot_mat" value="$(arg rho_rot_mat)" />
        <arg name="enable_rot_init" value="$(arg enable_rot_init)" />
        <arg name="rot_init_enable_gravity_prior" value="$(arg rot_init_enable_gravity_prior)" />
        <arg name="debug_rot_init_only" value="$(arg debug_rot_init_only)" />
        <arg name="eta_k" value="$(arg eta_k)" />
        <arg name="is_multi" value="true" />
        <arg name="drone_num" value="$(arg agent_num)" />
        <arg name="enable_linear_pose6d_solver" value="$(arg enable_linear_pose6d_solver)" />
        <arg name="ignore_infor" value="$(arg ignore_infor)" />
        <arg name="simulate_delay_ms" value="$(arg simulate_delay_ms)" />
        <arg name="max_solving_time" value="$(arg max_solving_time)" />
    </include>
    <node name="d2pgo" pkg="d2pgo" type="delay_launch_dpgo.sh" args="$(arg delay_launch)" if="$(arg wait_for_start)" />
    <node name="rvizvisualisation" pkg="rviz" type="rviz" output="log" args="-d $(find d2vins)/../config/d2pgo.rviz" if="$(arg rviz)" />
</launch>